### PR TITLE
Fix log record IDs

### LINKED ISSUE #28786: in warn macro the maxlog keyword is not effective
I'm trying reproducing the old behavior of `warn(msg, once = true)`. From the documentation I assumed that was done with the `@warn` macro and the `maxlog` keyword but it doesn't seem to work:

```julia
julia> function f(x)
           @warn "Test" maxlog=1
           return
       end
f (generic function with 1 method)

julia> f(1)
┌ Warning: Test
└ @ Main REPL[35]:2

julia> f(1)
┌ Warning: Test
└ @ Main REPL[35]:2
```

I'm on Julia 1.0:

```julia
julia> versioninfo()
Julia Version 1.0.0
Commit 5d4eaca0c9 (2018-08-08 20:58 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
```

### PR DESCRIPTION
**Pull request by adamslc**
This PR ensures that a given log statement will always produce the same ID. It is actually a bit tricky to implement because the ID should "uniquely identify the source location in the originating module, but should be stable across versions of the originating module, provided the log generating statement itself doesn't change" (according to the source).

This definitely needs some tests so that this doesn't break again.

Closes #28786.

### CODE REVIEW

**Comment by adamslc** [comment]
I think that this is ready to go now, but I do have a concern about the implementation as it stands now.

Suppose that some package has a function like this:
```julia
function f()
    for i in 1:1_000_000
        @info "Working"
    end
    @info "Done"
end
```

An application might want to filter out the "Working" logs, but pass along the "Done" log to the user. Prior to this PR, there was no good way to do that, but now the application can filter based on the log record ID, and everything works. Yay!

But now suppose that the package is modified to add a new log:
```julia
function f()
    @info "Starting"
    for i in 1:1_000_000
        @info "Working"
    end
    @info "Done"
end
```
Now the application will actually filter out the "Starting" log, and the user will be flooded with "Working" logs. One solution would be to include the message in the hash to compute the ID, but this would mean that the message (which may be expensive) would need to be computed before the `shouldlog` check.

**Comment by adamslc** [comment]
Bump. It would be great if this could make it in to v1.0.1.

**Comment by ViralBShah** [comment]
1.0.1 is frozen. Can go into 1.0.2

**Comment by oxinabox** [comment]
`id in values(_log_record_ids)`
is a $O(n)$ operation
when $n$ is the number of `_log_record_ids`
Probably fairly small.

Except if you do something like:

```
for ii in 1:10_000
     @info "round $ii  starting"
     #...
end
```

And I can see that code being written by a lot of people who don't know about logs now taking kwarguments, or even those who do, who think that this is prettier.

**Comment by adamslc** [comment]
I'm not really sure what you are arguing for? With this PR, every log record generated by your code with have the same ID, and the `id in values(_log_record_ids)` check will only happen on the first iteration.

Though the example you provide is another example of why including the message in the original id hash is a bad idea.

**Comment by adamslc** [comment]
Can this please get a backport label so it doesn't get lost?

**Comment by oxinabox** [comment]
> I'm not really sure what you are arguing for? 

I misread something in the code.

I wish I understood why this feature worked in 0.7-beta without having this code.

File: base/logging.jl [hunk 1]
```diff
@@ -311,12 +315,12 @@ function logmsg_code(_module, file, line, level, message, exs...)
             _module = $_module
             logger = current_logger_for_env(std_level, group, _module)
             if !(logger === nothing)
+                file = $file
```
Comment (KristofferC) [inline base/logging.jl]: Why were these moved?

Comment (adamslc) [reply base/logging.jl]: Those values are needed to compute `id`.

**Comment by c42f** [comment]
Oh, you're right there's a very serious problem with id's in 1.0. Not only are they not stable, but they're computed at every log invocation (very expensive, and will explode the log id storage!)  @oxinabox this was broken by the fix in f0e1d211c which went in very shortly before 1.0 and fixed a quite legitimate bug. Apparently my tests didn't cover this and I've been absent from the logging discussion for a few months, sorry (new job, etc etc...).

Hmm what to do. So I'm sure this PR is part of the solution, but I'm not sure it's quite right yet.

I'm thinking that automatically computed `id`s should *always* be statically computed, regardless of whether the user overrides the `_module`, `_file`, and `_line`. The heuristic we use may not suit in those cases anyway and the user can always override `_id` if they really must.

So what about if we go back to using the `_sourceinfo` for passing to `log_record_id` at compile time, with the autogenerated `_id` always arising from that?  This has predictable cost of only happening once at compile time, as the original implementation intended.

**Comment by c42f** [comment]
Thanks! I hope I'm back for a while... we'll see.

The memoization trick is a really clever and relatively nice way around the bootstrap problems with `basename`. We *could* use it for any metadata which is known (by definition or API contract) to be fixed for all invocations of the log statement, but which can't be computed at compile time. I don't think we've defined which of the special metadata keys are allowed to change between invocations of a given logging macro. One reasonable option might be "none" (in which case any of the special fields could be memoized). That might be too restrictive though.

Also regarding memoization, are there any problems with multithreading which could arise from that? Is it reasonable to assume pointer stores are atomic on all platforms we support?

**Comment by c42f** [comment]
Anyway, back to the issue at hand. For the particular case of `id`, we currently only promise in the docs that the `id` is an "arbitrary" but unique identifier for the log statement. So even a random uuid would be good enough to satisfy that, and much better than the current situation.

Thinking back, what I was trying to achieve with making `id` depend on overridable `_module` metadata was to allow wrapping logging macros inside other macros, having the somewhat-readable id respect the module where the outer macro was called. But this turns out not to work (unless you memoize it) because the inner logging macro only has syntax to work with, not the real `Module`.

Putting that all together, I think we should go back to simply hashing the (full) *syntax* of the log statement, combined with the module from the `_sourceinfo` and light deduplication. All of which is known and can be evaluated at compile time.

**Comment by c42f** [comment]
I've made a branch with my preferred fix (cjf/fix-logging-ids) I just haven't had the time to add additional tests.

**Comment by adamslc** [comment]
Feel free to take the tests from this PR.

**Comment by c42f** [comment]
Thanks, I cherry picked your commit containing the tests already :-)
